### DISCUSSION TITLE
Access nested components from rust?

### DISCUSSION THREAD

**Discussion by hellozyemlya** [slint-ui/slint#5794] [Q&A] [answered]
Window:
```
import { VerticalBox } from "std-widgets.slint";
import {MyTextComponent} from "child.slint";

export component AppWindow inherits Window {
    callback request-increase-value();
    VerticalBox {
        mytext := MyTextComponent {

        }
    }
}
```
Component:
```
import { VerticalBox } from "std-widgets.slint";

export component MyTextComponent inherits VerticalBox {
    in property <[{foo: string, col: color}]> model: [
        {foo: "abc", col: #f00 },
        {foo: "def", col: #00f },
    ];
    VerticalBox {
        VerticalLayout {
            for data in root.model: my-repeated-text := Text {
                color: data.col;
                text: data.foo;
            }
        }
    }
}
```
Is it possible somehow to access model property of MyTextComponent from instance of window? Something like `ui.mytext.model =` ...

**Comment by ogoffart** [comment]
You need to forward the property to the root, or use a global: https://releases.slint.dev/1.7.1/docs/slint/src/language/syntax/globals

**Reply by Enyium** [reply]
What's the preferred way of having multiple instances of complex widgets in one window that you don't want to completely abstract away towards Rust code (like you'd do with `ListView`s through their models), but instead keep awareness of the widget in Rust code? You could use a global for the various property bindings and callbacks of the widget. But should you then equip every property and callback with a widget index (wrapping everything in array models) and make every widget instance aware of its index? The widget instances wouldn't necessarily be rendered using `for`, but explicitly.

**Comment by ghost** [comment, marked as answer]
Hi,

there are two ways. One is to use a binding on the window the other is to use a global:

### Window binding

```slint
import { VerticalBox } from "std-widgets.slint";
import {MyTextComponent} from "child.slint";

export component AppWindow inherits Window {
    in property text-model <=> myteext.model;
    
    callback request-increase-value();
    VerticalBox {
        mytext := MyTextComponent {

        }
    }
}
```

```rust
let window = AppWindow::new().unwrap();

window.set_text_model(/* set your data here */);

window.run().unwrap();
```

### global

```slint
import { VerticalBox } from "std-widgets.slint";
import {MyTextComponent} from "child.slint";

export global Adapter {
   in property <[{foo: string, col: color}]> text_model;
}

export component AppWindow inherits Window {
    callback request-increase-value();
    VerticalBox {
        mytext := MyTextComponent {
             model: Adapter.text_model;
        }
    }
}
```

```rust
let window = AppWindow::new().unwrap();

Adapter::get(&window).set_text_model(/* set your data here */);

window.run().unwrap();
```

**Reply by hellozyemlya** [reply]
I see, so literally saying, only way to communicate with world is a property drilling. And if there is multiple nesting, it will require drill property all way down, right? Anyway, thanks for the answer.
